pkg/service: Handle leaked backends

aditighag opened the pull request:
The current logic to restore backends is brittle, and doesn't account for failure scenarios effectively. 

pkg/service: Handle duplicate backends
   
    In certain error scenarios, backends can be leaked, where
    they were deleted from the userspace state, but left in the
    datapath backends map. To reconcile datapath and userspace,
    identify such backends that were created with different IDs
    but same L3n4Addr hash.
    This commit builds up on previous commits that don't bail out
    on such error conditions (e.g., backend IDs mismatch during restore),
    and tracks backends that are currently referenced in service entries
    restored from the lb4_services map to restore backend entries.
    Furthermore, it uses the tracked state to delete any duplicate backends
    that were previously leaked.
    
    Fixes: b79a4a53 (pkg/service: Gracefully terminate service backends)

pkg/service: Restore services prior to backends

    The restore logic attempts to reconcile datapath state
    with the userspace post agent restart.
    Previously, it first restored backends from the `lb4_backends`
    map before restoring service entries from the `lb4_services`
    map. If there were error scenarios prior to agent restart (for
    example, backend map full because of leaked backends), the logic
    would fail to restore backends currently referenced in the services
    map (and as a result, selected for load-balancing traffic).
    
    This commit prioritizes restoring service entries followed by
    backend entries. Follow-up commit handles error cases such as leaked
    backends by keeping track of backends retrieved from restoration of
    service entries, and then using that to subsequently restore backends.

pkg/service: Don't bail out on failures
    
    The restore code attempts to reconcile datapath state with
    the userspace state post agent restart. Bailing out early
    on failures prevents any remediation from happening, so
    log any errors. Follow-up commits will try to handle leaked
    backends in the cluster if any.

```release-note
Handle leaked service backends that may lead to filling up of `lb4_backends` map and thereby connectivity issues.
```

Relates: https://github.com/cilium/cilium/issues/23551

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Handle duplicate backends

In certain error scenarios, backends can be leaked, where
they were deleted from the userspace state, but left in the
datapath backends map. To reconcile datapath and userspace,
identify such backends that were created with different IDs
but same L3n4Addr hash.
This commit builds up on previous commits that don't bail out
on such error conditions (e.g., backend IDs mismatch during restore),
and tracks backends that are currently referenced in service entries
restored from the lb4_services map to restore backend entries.
Furthermore, it uses the tracked state to delete any duplicate backends
that were previously leaked.

Fixes: b79a4a53 (pkg/service: Gracefully terminate service backends)

Signed-off-by: Aditi Ghag <aditi@cilium.io>

tommyp1ckles commented on pkg/service/service.go:
```diff
@@ -1402,6 +1407,30 @@ func (s *Service) deleteOrphanBackends() error {
 
+// Check for duplicate backends that have same address but different ids
+func (s *Service) deleteDuplicateBackends() error {
+	backends, err := s.lbmap.DumpBackendMaps()
```
Could we use DumpWithCallback instead to avoid the extra copying of the map into an intermediate slice?

aditighag replied:
Good idea! In fact, we just need IDs, so we can skip copying backend values. Thanks!
Edit: On second thought, we may not need another iteration over the backends map.

aditighag replied:
This code is outdated. Please check the latest changes.

aditighag committed:
pkg/service: Don't bail out on failures

The restore code attempts to reconcile datapath state with
the userspace state post agent restart. Bailing out early
on failures prevents any remediation from happening, so
log any errors. Follow-up commits will try to handle leaked
backends in the cluster if any.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Restore services prior to backends

The restore logic attempts to reconcile datapath state
with the userspace post agent restart.
Previously, it first restored backends from the `lb4_backends`
map before restoring service entries from the `lb4_services`
map. If there were error scenarios prior to agent restart (for
example, backend map full because of leaked backends), the logic
would fail to restore backends currently referenced in the services
map (and as a result, selected for load-balancing traffic).

This commit prioritizes restoring service entries followed by
backend entries. Follow-up commit handles error cases such as leaked
backends by keeping track of backends retrieved from restoration of
service entries, and then using that to subsequently restore backends.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

tommyp1ckles commented on pkg/service/service.go:
```diff
@@ -240,6 +240,7 @@ type Service struct {
 
 	backendRefCount counter.StringCounter
 	backendByHash   map[string]*lb.Backend
+	backendsById    map[lb.BackendID]struct{}
```
Is this only used one for restore, if so can we null this out after the restore step is complete so it can be GCd?

aditighag replied:
It's not used anymore. That's why I ended up adding a local variable. I'll remove the unused variable.

tommyp1ckles commented on pkg/service/service.go:
```diff
@@ -953,20 +955,28 @@ func (s *Service) GetDeepCopyServicesByName(name, namespace string) (svcs []*lb.
-		return err
+	// Restore service cache from BPF maps
+	if err := s.restoreServicesLocked(backendsById); err != nil {
+		errs = multierr.Append(errs,
```
nit: I don't think this ever actually returns a non-nil error value.

aditighag replied:
Sorry, I don't understand. Can you elaborate? The zero value of errs is `nil`.

tommyp1ckles replied:
sorry should've specified the actual line, restoreServicesLocked doesn't ever appear to return anything but nil for its error.

tommyp1ckles commented on pkg/service/service.go:
```diff
@@ -953,20 +955,28 @@ func (s *Service) GetDeepCopyServicesByName(name, namespace string) (svcs []*lb.
-	if err := s.restoreServicesLocked(); err != nil {
-		return err
+	// Restore backend IDs
+	if err := s.restoreBackendsLocked(backendsById); err != nil {
```
I think restoreServices will do nothing if the map iteration fails, in that case backendsByID is never populated.

In that case, restoreBackends will delete all backends.   Could that cause any problems?

aditighag replied:
So in `restoreBackendsLocked`, the logic that handles duplicate backends was missing a check for identifying "duplicate" backends: backends that are found by hash, but not by id. `backendRefCount` is populated when services are restored.

```
-                if _, ok := svcBackendsById[b.ID]; !ok {
+               if _, ok := svcBackendsById[b.ID]; !ok && s.backendRefCount[b.L3n4Addr.Hash()] != 0 {
```

 I've updated this code locally, so the second check should return false. Hope that makes sense.

tommyp1ckles replied:
I think what I meant was that because the map is only populated in restoreServicesLocked, if that [fails for some reason](https://github.com/cilium/cilium/pull/24681/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R1452) then the map will just remain empty, in which case every backend will be considered a duplicate and be removed: https://github.com/cilium/cilium/pull/24681/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R1379-R1407

Not sure if this is actually a problem.

tommyp1ckles replied:
I guess the only thing I could think of happening is say theres some transient error with the iterate map call.

In that case, all backends could get removed, which might cause some issue with existing connections while services get restored from k8s apiserver?

aditighag replied:
Synced up offline.

brb commented on pkg/service/service.go:
```diff
@@ -1369,6 +1374,35 @@ func (s *Service) restoreBackendsLocked() error {
 			logfields.BackendState:     b.State,
 			logfields.BackendPreferred: b.Preferred,
 		}).Debug("Restoring backend")
+		if _, ok := svcBackendsById[b.ID]; !ok && s.backendRefCount[b.L3n4Addr.Hash()] != 0 {
```
@aditighag Too late to the party, but a few questions:
* Isn't the `s.backendRefCount[b.L3n4Addr.Hash()] != 0` check redundant, as the refcount of such a backend can be incremented only if it belongs to a service (thus, `!ok` would evaluate to false).
* Do I read it correctly that backends in the `Terminating` state would be removed before the grace period happens? I think it's not a big deal, as cilium-agent restarts do not expect to happen very often. Just what happens if we receive an EndpointSlice delete event, and cilium-agent doesn't have any reference to previously `Terminating` backends?

brb replied:
Another thing it would be helpful if you could elaborate on (e.g., in a commit msg) under what circumstances the backend leaks happened.
